subsystems/networking/bluetooth.go

```diff
@@ -0,0 +1,172 @@
+					}
+					if psk == "" {
+						psk, pskErr = bwp.readPsk()
+						if pskErr != nil && !errors.As(pskErr, emptyBluetoothCharacteristicError{}) {
+							return
+						}
+					}
+				}
+				if requiresCloudCredentials {
+					if robotPartKeyID == "" {
+						robotPartKeyID, robotPartKeyIDErr = bwp.readRobotPartKeyID()
+						if robotPartKeyIDErr != nil && !errors.As(robotPartKeyIDErr, emptyBluetoothCharacteristicError{}) {
+							return
+						}
+					}
+					if robotPartKey == "" {
+						robotPartKey, robotPartKeyErr = bwp.readRobotPartKey()
+						if robotPartKeyErr != nil && !errors.As(robotPartKeyErr, emptyBluetoothCharacteristicError{}) {
+							return
+						}
+					}
+				}
+				if requiresWiFiCredentials && requiresCloudCredentials &&
+					ssid != "" && psk != "" && robotPartKeyID != "" && robotPartKey != "" {
+					return
+				} else if requiresWiFiCredentials && ssid != "" && psk != "" {
+					return
+				} else if requiresCloudCredentials && robotPartKeyID != "" && robotPartKey != "" {
+					return
+				}
+
+				// Not ready to return (do not have the minimum required set of credentials), so sleep and try again.
+				time.Sleep(time.Second)
+			}
+		}
+
+	}, wg.Done)
+
+	wg.Wait()
+
+	return &userInput{
+		SSID: ssid, PSK: psk, PartID: robotPartKeyID, Secret: robotPartKey,
+	}, errors.Join(ctxErr, ssidErr, pskErr, robotPartKeyIDErr, robotPartKeyErr)
+}
+
+/** Helper methods for low-level system calls and read/write requests to/from bluetooth characteristics **/
+
+func (bwp *bluetoothServiceLinux) enableAutoAcceptPairRequest() {
+	// TODO APP-7655: Implement method to auto-accept pairing requests to the BLE peripheral.
+}
+
+func (bwp *bluetoothServiceLinux) readSsid() (string, error) {
+	return "", errors.New("TODO APP-7653: Implement helper methods to read SSID, passkey, robot part key ID, and robot part key" +
+		" values from BLE peripheral characteristics")
+}
+
+func (bwp *bluetoothServiceLinux) readPsk() (string, error) {
+	return "", errors.New("TODO APP-7653: Implement helper methods to read SSID, passkey, robot part key ID, and robot part key" +
+		" values from BLE peripheral characteristics")
+}
+
+func (bwp *bluetoothServiceLinux) readRobotPartKeyID() (string, error) {
+	return "", errors.New("TODO APP-7653: Implement helper methods to read SSID, passkey, robot part key ID, and robot part key" +
+		" values from BLE peripheral characteristics")
+}
+
+func (bwp *bluetoothServiceLinux) readRobotPartKey() (string, error) {
+	return "", errors.New("TODO APP-7653: Implement helper methods to read SSID, passkey, robot part key ID, and robot part key" +
+		" values from BLE peripheral characteristics")
+}
+
+// NewBluetoothWiFiProvisioningServiceLinux returns a service which accepts credentials over bluetooth to provision a robot and its WiFi connection.
+func NewBluetoothWiFiProvisioningService(ctx context.Context, logger logging.Logger, name string) (*bluetoothServiceLinux, error) {
+	switch os := runtime.GOOS; os {
```

Comment:
Generally we don't want to do per-OS logic at runtime. Go has a pretty clever mechanism (build constraints) for making different versions of a file (and therefore the functions it contains) build only under certain conditions. E.g. the "windows" build would include only the windows code, and the regular/linux file would include the linux versions. That keeps code cleaner and resulting binaries smaller as well.

See https://pkg.go.dev/cmd/go#hdr-Build_constraints for details, but feel free to ask me for more examples as we've used it quite a bit in RDK itself.